### DISCUSSION TITLE
Cache data on CPU memory and then perform non-deterministic GPU transforms

### DISCUSSION THREAD

**Discussion by hw-ju** [Project-MONAI/tutorials#1408] [Q&A]
Hi! 
In [this tutorial for using GPU transform](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb), the intermediate data  (i.e., after going through a chain of deterministic transforms) is cached on **GPU** and then non-deterministic **GPU** transform is applied before feeding data into the model.

Can we cache intermediate data on **CPU** memory, and then perform non-deterministic **GPU** transforms? If yes, how to do it?

**Comment by KumoLiu** [comment]
Hi @hw-ju, yes you can. Just remove `EnsureType`. I think regular training in [this tutorial](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb) cache data in CPU instead of GPU.
Thanks!

**Reply by hw-ju** [reply]
Hi @KumoLiu! Thanks for your prompt reply! I have a follow-up question.
Let's say we're using `CacheDataset` with a transform chain like 
```
[deterministic_CPU_transforms, EnsureType, randomized_transform (e.g., RandCropByPosNegLabeld)]
```
where `randomized_transform` is performed on GPU.

After I remove `EnsureType`, it becomes 
```
[deterministic_CPU_transforms, randomized_transform (e.g., RandCropByPosNegLabeld)]
```
where `randomized_transform` will be run on CPU, correct? How can I tell this `randomized_transform` (e.g., RandCropByPosNegLabeld) to run on GPU?

**Reply by KumoLiu** [reply]
Yes, you are right, just remove `EnsureType`, no more things need to do since data are already on the CPU. If you want `randomized_transform` to run on GPU, just add `EnsureType`. Thanks!

**Reply by hw-ju** [reply]
@KumoLiu 
With `EnsureType`, it caches on **GPU** and run `randomized_transform` on **GPU**.
Without `EnsureType`, it caches on **CPU** and run `randomized_transform` on **CPU**.
Is there any way to cache on **CPU** and run `randomized_transform` on **GPU**?

**Reply by KumoLiu** [reply]
I'm afraid not.
Perhaps `EnsureType` can have a flag to indicate whether to cache it. 
cc @wyli

**Reply by KumoLiu** [reply]
I just noticed that you can use `PersistentDataset` with `EnsureType`.

**Reply by hw-ju** [reply]
@KumoLiu 
If I use `PersistentDataset` with `EnsureType`, then it will cache on some **persistent storage** and will perform `randomized_transform` on **GPU**, corrrect?

**Reply by KumoLiu** [reply]
Yes, just put `EnsureType` before `randomized_transform`. Thanks!

**Reply by hw-ju** [reply]
@KumoLiu Thank you so much for helping me so patiently :)!!
